Support dual nics

KimForss opened the pull request:
## Problem
Customers need the ablility to provide two IP Addresses for each VM 

## Solution
Add a second network interface card (optionally) The ability to set the NIC should be per tier, ie per app and per db. The second nic should be attached to the admin subnet.

## Tests
add    "dual_nics": true in either the application or the database block in the json file

## Notes
<Additional comments for the PR>

rukawata commented on deploy/terraform/terraform-units/modules/sap_system/app_tier/infrastructure.tf:
```diff
@@ -1,3 +1,20 @@
+# Creates admin subnet of SAP VNET
+resource "azurerm_subnet" "sap-admin" {
```
Do we need one additional subnet per app/web/scs? Or one for all? Do you have any reference doc?
And again, just as db, i think admin subnet is the one exists by default, and db subnet for db, app subnet for app, which would be used for internal traffic.

KimForss replied:
One subnet for all admin nics and one for app, one for db and an optional for web

rukawata reviewed (commented):
Thanks! Some comments.

KimForss commented on deploy/terraform/terraform-units/modules/sap_system/anydb_node/variables_local.tf:
```diff
@@ -71,17 +84,20 @@ locals {
+  anydb_version  = try(local.anydb.db_version, "")
+
+  // Dual network cards
+  use_two_network_cards = try(local.anydb.dual_nics, false)
```
done

KimForss commented on deploy/terraform/terraform-units/modules/sap_system/anydb_node/infrastructure.tf:
```diff
@@ -56,6 +56,25 @@ resource "azurerm_availability_set" "anydb" {
+# VNET ================================================================================================
+
+# Creates admin subnet of SAP VNET
+resource "azurerm_subnet" "anydb-admin" {
```
The default for all my customers except Achmea is one nic on the db attached to the db subnet and the optional second would be connected to the admin subnet

rukawata reviewed (commented):
There may be some problem in the way admin subnet is created and used. Please check the comments. Thanks!

KimForss commented on deploy/terraform/terraform-units/modules/sap_system/anydb_node/vm-anydb.tf:
```diff
@@ -24,6 +24,29 @@ resource "azurerm_network_interface" "anydb" {
 }
 
+# Creates the DB traffic NIC and private IP address for database nodes
+resource "azurerm_network_interface" "anydb-admin" {
```
changed

KimForss commented on deploy/terraform/terraform-units/modules/sap_system/anydb_node/vm-anydb.tf:
```diff
@@ -40,13 +63,14 @@ resource "azurerm_linux_virtual_machine" "dbserver" {
+
+  network_interface_ids = local.anydb_dual_nics ? (
+    [azurerm_network_interface.anydb-admin[count.index].id, azurerm_network_interface.anydb[count.index].id]) : (
+    [azurerm_network_interface.anydb[count.index].id]
```
done

KimForss commented on deploy/terraform/terraform-units/modules/sap_system/anydb_node/vm-anydb.tf:
```diff
@@ -40,13 +63,14 @@ resource "azurerm_linux_virtual_machine" "dbserver" {
-  zone = local.zonal_deployment ? (
-    local.db_server_count == local.db_zone_count ? local.zones[count.index % local.db_zone_count] : null) : (
-    null
+  zone = local.zonal_deployment && (local.db_server_count == local.db_zone_count) ? local.zones[count.index % local.db_zone_count] : null
```
merge conflict/rebase

rukawata replied:
This comes from a beta2 merge (check [here](https://github.com/Azure/sap-hana/commit/4ac19df3358e1ed5b0aca141e3282be6c8a7a7f6)).
It seems difficult to just revert - maybe just make a new commit change it back.

KimForss commented on deploy/terraform/terraform-units/modules/sap_system/anydb_node/vm-anydb.tf:
```diff
@@ -108,13 +132,13 @@ resource "azurerm_windows_virtual_machine" "dbserver" {
-    local.db_server_count == local.db_zone_count ? local.zones[count.index % local.db_zone_count] : null) : (
-    null
-  )
+  zone = local.zonal_deployment && (local.db_server_count == local.db_zone_count) ? local.zones[count.index % local.db_zone_count] : null
```
merge conflict/rebase

KimForss commented on deploy/terraform/terraform-units/modules/sap_system/app_tier/infrastructure.tf:
```diff
@@ -1,3 +1,20 @@
+
+# Imports data of existing SAP admin subnet
+data "azurerm_subnet" "sap-admin" {
+  count                = local.enable_deployment  && local.apptier_dual_nics ? (local.sub_admin_exists ? 1 : 0) : 0
```
move to common resources

KimForss commented on deploy/terraform/terraform-units/modules/sap_system/anydb_node/variables_local.tf:
```diff
@@ -55,6 +55,19 @@ locals {
   vnet_sap_name_prefix = try(substr(upper(local.vnet_sap_name), -5, 5), "") == "-VNET" ? split("-", local.vnet_sap_name)[(local.vnet_nr_parts - 2)] : local.vnet_sap_name
 
+  // Admin subnet
+  var_sub_admin    = try(var.infrastructure.vnets.sap.subnet_admin, {})
```
moved it to common resources

KimForss commented on deploy/terraform/terraform-units/modules/sap_system/app_tier/vm-scs.tf:
```diff
@@ -7,9 +7,37 @@ resource "azurerm_network_interface" "scs" {
+
+# Create Admin NICs
+resource "azurerm_network_interface" "scs-admin" {
+  count                         = local.enable_deployment && local.apptier_dual_nics ? (local.scs_high_availability ? 2 : 1) : 0
```
fixed

KimForss commented on deploy/terraform/terraform-units/modules/sap_system/app_tier/vm-app.tf:
```diff
@@ -7,9 +7,37 @@ resource "azurerm_network_interface" "app" {
+      cidrhost(local.sub_admin_exists ?
+        data.azurerm_subnet.sap-admin[0].address_prefixes[0] :
+        azurerm_subnet.sap-admin[0].address_prefixes[0],
+        tonumber(count.index) + 15
```
I agree - this is a big challenge as the current model will fail for the second SDU deployment into the same subnets

can we do that in a separate PR as the version I need now is with fixed IP addresses

KimForss commented on deploy/terraform/terraform-units/modules/sap_system/common_infrastructure/iscsi.tf:
```diff
@@ -56,7 +56,7 @@ resource "azurerm_network_security_rule" "iscsi" {
   destination_port_range       = "*"
   source_address_prefix        = "*"
-  destination_address_prefixes = try(var.subnet-sap-admin.address_prefixes, "*")
+  destination_address_prefixes = try(local.sub_admin_exists ? data.azurerm_subnet.admin[0].address_prefixes : azurerm_subnet.admin[0].address_prefixes, "*")
```
removed

YunziZhangAzure commented on deploy/terraform/terraform-units/modules/sap_system/app_tier/vm-webdisp.tf:
```diff
@@ -94,9 +116,11 @@ resource "azurerm_windows_virtual_machine" "web" {
+  network_interface_ids = local.apptier_dual_nics ? (
+    [azurerm_network_interface.web-admin[count.index].id, azurerm_network_interface.web[count.index].id]) : (
+    [azurerm_network_interface.web[count.index].id]
+  )
```
Maybe the format can be better

YunziZhangAzure commented on deploy/terraform/terraform-units/modules/sap_system/anydb_node/variables_local.tf:
```diff
@@ -55,6 +59,19 @@ locals {
+  var_sub_admin_nsg    = try(var.infrastructure.vnets.sap.subnet_admin.nsg, {})
+  sub_admin_nsg_arm_id = try(local.var_sub_admin_nsg.arm_id, "")
+  sub_admin_nsg_exists = length(local.sub_admin_nsg_arm_id) > 0 ? true : false
+  sub_admin_nsg_name   = local.sub_admin_nsg_exists ? try(split("/", local.sub_admin_nsg_arm_id)[8], "") : try(local.var_sub_admin_nsg.name, format("%s%s", local.prefix, local.resource_suffixes.admin-subnet-nsg))
```
Since the admin_subnet is generated in the common_infrastructure and passed to anydb module, there's no need to deduct all these variables again(from line 63 to line 73).
e.g: 
var_sub_admin = var.admin_subnet
sub_admin_arm_id = var_sub_admin.id
...

The nsg of admin can be passed to this child module(anydb) as well, like admin_subnet.

Generally speaking, in the "child" module(anydb, hdb, app tier), I think it is better to utilize the admin_subnet which are passed from the common_infrastructure, instead of retrieving these variables from input json again.

KimForss replied:
I agree - but I was thinking of removing it once we decouple the network creation

YunziZhangAzure commented on deploy/terraform/terraform-units/modules/sap_system/app_tier/variables_local.tf:
```diff
@@ -70,6 +75,19 @@ locals {
+  var_sub_admin_nsg    = try(var.infrastructure.vnets.sap.subnet_admin.nsg, {})
+  sub_admin_nsg_arm_id = try(local.var_sub_admin_nsg.arm_id, "")
+  sub_admin_nsg_exists = length(local.sub_admin_nsg_arm_id) > 0 ? true : false
+  sub_admin_nsg_name   = local.sub_admin_nsg_exists ? try(split("/", local.sub_admin_nsg_arm_id)[8], "") : try(local.var_sub_admin_nsg.name, format("%s%s", local.prefix, local.resource_suffixes.admin-subnet-nsg))
```
In the "child" module(anydb, hdb, app tier), I think it is better to utilize the admin_subnet which are passed from the common_infrastructure, instead of retrieving these variables from input json again.

KimForss replied:
I agree - but I was thinking of removing it once we decouple the network creation

YunziZhangAzure commented on deploy/terraform/terraform-units/modules/sap_system/anydb_node/vm-anydb.tf:
```diff
@@ -13,13 +13,28 @@ resource "azurerm_network_interface" "anydb" {
-        azurerm_subnet.anydb[0].address_prefixes[0])
-      ), tonumber(count.index) + 10)
-    )
+    private_ip_address = try(local.anydb_vms[count.index].db_nic_ip, cidrhost(local.sub_db_exists ? (
```
I'm afraid the code here may not be correct. 
anydb_vms[count.index].db_nic_ip has been given a default value "null" as default in varible_local.tf, so this "try" expression never hits the second part " cidrhost(local.sub_db_exist....". 

However, the logic should be: if the db nic ip is not specified in the input json, an ip with anydb subnet address prefix should be generated by calling this cidrhost

Can you try running some tests on this part to make sure the code here is correct?

KimForss replied:
It was not correct - fixing it

YunziZhangAzure commented on deploy/terraform/terraform-units/modules/sap_system/common_infrastructure/infrastructure.tf:
```diff
@@ -35,6 +35,46 @@ data "azurerm_virtual_network" "vnet-sap" {
+
+# Creates SAP admin subnet nsg
+resource "azurerm_network_security_group" "admin" {
+  count               = local.enable_deployment ? (local.sub_admin_nsg_exists ? 0 : 1) : 0
```
```suggestion
  count               = (local.enable_deployment && ! local.sub_admin_nsg_exists) ? 1 : 0
```

YunziZhangAzure commented on deploy/terraform/terraform-units/modules/sap_system/common_infrastructure/infrastructure.tf:
```diff
@@ -35,6 +35,46 @@ data "azurerm_virtual_network" "vnet-sap" {
+
+# Imports the SAP admin subnet nsg data
+data "azurerm_network_security_group" "admin" {
+  count               = local.enable_deployment ? (local.sub_admin_nsg_exists ? 1 : 0) : 0
```
```suggestion
count               = (local.enable_deployment && local.sub_admin_nsg_exists) ? 1 : 0
```

YunziZhangAzure commented on deploy/terraform/terraform-units/modules/sap_system/app_tier/vm-scs.tf:
```diff
@@ -7,9 +7,35 @@ resource "azurerm_network_interface" "scs" {
+  }
+}
+
+# Create Admin NICs
```
Please use // when commenting.

YunziZhangAzure commented on deploy/terraform/terraform-units/modules/sap_system/app_tier/vm-app.tf:
```diff
@@ -7,9 +7,35 @@ resource "azurerm_network_interface" "app" {
+    name      = "IPConfig1"
+    subnet_id = local.sub_app_exists ? data.azurerm_subnet.subnet-sap-app[0].id : azurerm_subnet.subnet-sap-app[0].id
+    private_ip_address = try(local.app_nic_ips[count.index],
+      cidrhost(local.sub_web_exists ?
```
Why do we need to change the logic here?

KimForss replied:
copy paste error

YunziZhangAzure reviewed (commented):
Thanks for addressing comments! Overall looks good.

Just a reminder, the admin nic ip may need to be output from anydb/app_tier module and passed to output_files module to generate hosts.yml for Ansible use in the future, but I guess this doesn't need to be implemented in this pr, just want to bring it up so we can be aware of this potential requirement.

KimForss commented on deploy/terraform/terraform-units/modules/sap_system/common_infrastructure/infrastructure.tf:
```diff
@@ -35,6 +35,46 @@ data "azurerm_virtual_network" "vnet-sap" {
+
+# Imports the SAP admin subnet nsg data
+data "azurerm_network_security_group" "admin" {
+count               = (local.enable_deployment && local.sub_admin_nsg_exists) ? 1 : 0
```
done

rukawata commented on deploy/terraform/terraform-units/modules/sap_system/anydb_node/vm-anydb.tf:
```diff
@@ -13,13 +13,35 @@ resource "azurerm_network_interface" "anydb" {
+  }
+}
+
+# Creates the DB traffic NIC and private IP address for database nodes
```
The comment is incorrect, this is to create admin traffic NIC rather than DB.

KimForss replied:
fixed

rukawata commented on deploy/terraform/terraform-units/modules/sap_system/common_infrastructure/infrastructure.tf:
```diff
@@ -35,6 +35,46 @@ data "azurerm_virtual_network" "vnet-sap" {
 
+// Creates admin subnet of SAP VNET
+resource "azurerm_subnet" "admin" {
+  count                = local.sub_admin_exists ? 0 : 1
```
In the old behavior, we only create admin subnet when hdb deployment is enabled.
So to be consistent, we should only create admin subnet when:
- hdb deployed or
- anydb dual nics enabled or
- app tier dual nics enabled
Otherwise, admin subnet should not be created.

KimForss replied:
fixed

rukawata reviewed (commented):
Thanks for moving the admin subnet to infra - this makes more sense.
Some questions and concerns, please check.

KimForss commented on deploy/terraform/run/sap_system/output.tf:
```diff
@@ -0,0 +1,13 @@
+// Output IP Addresses
+
+output "anydb_admin_ip" {
+  value = azurerm_network_interface.anydb_admin.*.ip_configuration.private_ip_address
+}
+
+output "anydb_db_ip" {
+  value = azurerm_network_interface.anydb_db.*.ip_configuration.private_ip_address
+}
+
+output "anydb_lb_ip" {
+  value = azurerm_lb.anydb.*.frontend_ip_configuration.private_ip_address
+}
```
This is for a future PR - so that we can integrate with DNS - will remove them for now

KimForss commented on deploy/terraform/run/sap_system/module.tf:
```diff
@@ -154,6 +156,7 @@ module "anydb_node" {
   ppg                        = module.common_infrastructure.ppg
   naming                     = module.sap_namegenerator.naming
   custom_disk_sizes_filename = var.db_disk_sizes_filename
+  admin_subnet               = module.common_infrastructure.admin_subnet
```
Done

KimForss commented:
> Thanks for addressing comments! Overall looks good.
> 
> Just a reminder, the admin nic ip may need to be output from anydb/app_tier module and passed to output_files module to generate hosts.yml for Ansible use in the future, but I guess this doesn't need to be implemented in this pr, just want to bring it up so we can be aware of this potential requirement.

Fixed

rukawata commented on deploy/terraform/terraform-units/modules/sap_system/app_tier/outputs.tf:
```diff
@@ -1,11 +1,60 @@
+output "nics_web_admin" {
+  value = azurerm_network_interface.web_admin
+}
+
+
```
nit:
```suggestion

```

rukawata commented on deploy/terraform/terraform-units/modules/sap_system/common_infrastructure/infrastructure.tf:
```diff
@@ -35,6 +35,46 @@ data "azurerm_virtual_network" "vnet-sap" {
 
+// Creates admin subnet of SAP VNET
+resource "azurerm_subnet" "admin" {
+  count                = local.sub_admin_exists ? 0 : local.create_admin_subnet ? 1 : 0
```
```suggestion
  count                = ! local.sub_admin_exists && local.create_admin_subnet ? 1 : 0
```

rukawata commented on deploy/terraform/terraform-units/modules/sap_system/common_infrastructure/infrastructure.tf:
```diff
@@ -35,6 +35,46 @@ data "azurerm_virtual_network" "vnet-sap" {
+
+# Imports data of existing SAP admin subnet
+data "azurerm_subnet" "admin" {
+  count                = local.sub_admin_exists ? 1 : 0
```
only import existing admin subnet when needed:
```suggestion
  count                = local.sub_admin_exists  && local.create_admin_subnet ? 1 : 0
```

rukawata commented on deploy/terraform/terraform-units/modules/sap_system/common_infrastructure/infrastructure.tf:
```diff
@@ -35,6 +35,46 @@ data "azurerm_virtual_network" "vnet-sap" {
+
+# Creates SAP admin subnet nsg
+resource "azurerm_network_security_group" "admin" {
+  count               = ! local.sub_admin_nsg_exists ? 1 : 0
```
Only create NSG for admin subnet when it is created/imported.
```suggestion
  count               = ! local.sub_admin_nsg_exists  && local.create_admin_subnet ? 1 : 0
```

KimForss replied:
fixed

rukawata commented on deploy/terraform/terraform-units/modules/sap_system/common_infrastructure/infrastructure.tf:
```diff
@@ -35,6 +35,46 @@ data "azurerm_virtual_network" "vnet-sap" {
+
+# Imports the SAP admin subnet nsg data
+data "azurerm_network_security_group" "admin" {
+  count               = local.sub_admin_nsg_exists ? 1 : 0
```
same as above.

KimForss replied:
fixed

rukawata commented on deploy/terraform/terraform-units/modules/sap_system/common_infrastructure/infrastructure.tf:
```diff
@@ -35,6 +35,46 @@ data "azurerm_virtual_network" "vnet-sap" {
+
+# Associates SAP admin nsg to SAP admin subnet
+resource "azurerm_subnet_network_security_group_association" "Associate-admin" {
+  count                     = (signum((local.sub_admin_exists ? 0 : 1) + (local.sub_admin_nsg_exists ? 0 : 1))) 
```
same as above.

KimForss replied:
fixed

rukawata commented on deploy/terraform/terraform-units/modules/sap_system/common_infrastructure/variables_local.tf:
```diff
@@ -88,6 +84,8 @@ locals {
   //iSCSI
   var_iscsi = try(local.var_infra.iscsi, {})
 
+  create_admin_subnet = try(var.application.dual_nics, false) || try(var.databases[0].dual_nics, false) || length(local.hana-databases) > 0
```
considering the fact we also import subnet, like `enable_admin_subnet` or some others may make more sense?

KimForss replied:
yes, fixed

rukawata commented on deploy/terraform/terraform-units/modules/sap_system/output_files/variables_local.tf:
```diff
@@ -62,23 +62,42 @@ variable "hana-database-info" {
+variable "nics_anydb_admin" {
+  description = "List of Admin NICs for the anyDB VMs"
+}
+
+
+
```
nit:
```suggestion

```

rukawata commented on deploy/terraform/terraform-units/modules/sap_system/output_files/variables_local.tf:
```diff
@@ -128,11 +147,11 @@ locals {
-  ips-app = [for key, value in var.nics-app : value.private_ip_address]
-  ips-web = [for key, value in var.nics-web : value.private_ip_address]
+  ips-scs = [for key, value in var.nics_scs : value.private_ip_address]
+  ips-app = [for key, value in var.nics_app : value.private_ip_address]
+  ips-web = [for key, value in var.nics_web : value.private_ip_address]
 
-  ips-anydbnodes = [for key, value in var.nics-anydb : value.private_ip_address]
+  ips-anydbnodes = [for key, value in var.nics_anydb : value.private_ip_address]
```
This will be the ip in host and host.yml, meaning needs to be the ip that can be accessed from deployer.
So if no admin subnet, it is correct, otherwise, need to pass in the admin nic ips.

KimForss replied:
fixed

rukawata reviewed (commented):
Thanks for addressing the comments so quickly. Some additional concerns.

rukawata commented on deploy/terraform/terraform-units/modules/sap_system/app_tier/outputs.tf:
```diff
@@ -1,11 +1,58 @@
+}
+
+output "app_admin_ip" {
+  value = azurerm_network_interface.app-admin.*.ip_configuration.private_ip_address
```
```suggestion
  value = azurerm_network_interface.app_admin.*.ip_configuration.private_ip_address
```
There are other occurrence, please check log [here](https://dev.azure.com/azuresaphana/Azure-SAP-HANA/_build/results?buildId=13528&view=logs&jobId=ed898bca-1be4-5573-ffb5-764df6a725fe&j=ed898bca-1be4-5573-ffb5-764df6a725fe&t=3515a1e6-b72c-5f3d-99f5-7289c6342fae).

rukawata reviewed (approved):
Thanks! Now looks good :)

YunziZhangAzure commented on deploy/terraform/terraform-units/modules/sap_system/anydb_node/outputs.tf:
```diff
@@ -1,5 +1,22 @@
+
+
+output "anydb_admin_ip" {
+  value = azurerm_network_interface.anydb_admin.*.ip_configuration.private_ip_address
```
*. without [] seems to be a legacy splat expression. 

https://www.terraform.io/docs/configuration/expressions.html

It looks like azurerm_network_interface.anydb_admin[*].ip_configuration.private_ip_address is more recommended by Terraform. 

But I think the current expression can work, so it's up to you to decide which expression you want to use. :)

KimForss commented:
Let’s keep it like this and revisit when we add DNS support

________________________________
From: Yunzi Zhang <notifications@github.com>
Sent: Thursday, October 29, 2020 10:42:12 PM
To: Azure/sap-hana <sap-hana@noreply.github.com>
Cc: Kimmo Forss <Kimmo.Forss@microsoft.com>; Assign <assign@noreply.github.com>
Subject: Re: [Azure/sap-hana] Support dual nics (#822)


@YunziZhangAzure commented on this pull request.

________________________________

In deploy/terraform/terraform-units/modules/sap_system/anydb_node/outputs.tf<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fsap-hana%2Fpull%2F822%23discussion_r514553608&data=04%7C01%7Ckimmo.forss%40microsoft.com%7C29d53a2378814b7a9e9508d87c4b1d75%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637396009344990347%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=ZsKdDZsEVUhrgQheBs%2FboxhgLOwrsSnFX%2FjwSE7p7nI%3D&reserved=0>:

> @@ -1,5 +1,22 @@
-output "nics-anydb" {
-  value = azurerm_network_interface.anydb
+output "nics_anydb" {
+  value = azurerm_network_interface.anydb_db
+}
+
+output "nics_anydb_admin" {
+  value = azurerm_network_interface.anydb_admin
+}
+
+
+output "anydb_admin_ip" {
+  value = azurerm_network_interface.anydb_admin.*.ip_configuration.private_ip_address


*. without [] seems to be a legacy splat expression.

https://www.terraform.io/docs/configuration/expressions.html<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwww.terraform.io%2Fdocs%2Fconfiguration%2Fexpressions.html&data=04%7C01%7Ckimmo.forss%40microsoft.com%7C29d53a2378814b7a9e9508d87c4b1d75%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637396009345000341%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=8gya8m%2Fjkxi%2F6dKZu2Vj4%2BlHKIub6mDG7azcpVj6l00%3D&reserved=0>

It looks like azurerm_network_interface.anydb_admin[*].ip_configuration.private_ip_address is more recommended by Terraform.

But I think the current expression can work, so it's up to you to decide which expression you want to use. :)

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fsap-hana%2Fpull%2F822%23pullrequestreview-520064416&data=04%7C01%7Ckimmo.forss%40microsoft.com%7C29d53a2378814b7a9e9508d87c4b1d75%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637396009345000341%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=yZVdOV3NyOXYG%2Fr0ONt8shGFM47KgDINMPRvQ3bzFjo%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAGCOMTERYLWQEV5G3LOTVJDSNHHSJANCNFSM4SGPC7XA&data=04%7C01%7Ckimmo.forss%40microsoft.com%7C29d53a2378814b7a9e9508d87c4b1d75%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637396009345010335%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=afrC9fm3UH1qUzMBFGWy0A1WCQVLk4Gw1%2BNtDA12LdM%3D&reserved=0>.
